### PR TITLE
Smart contract to verify an address and for accept an invitation

### PR DESCRIPTION
**Pull request by andresaiello**
### Summary
- Register contract to verify an address
- Reference contract to accept a referral code

### CODE REVIEW

**Comment by nautel2000** [comment]
passed

File: packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    function _verifySignature(address inviter, Signature calldata signature) private pure {
+        bytes32 payloadHash = keccak256(abi.encode(inviter));
+        bytes32 messageHash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", payloadHash));
+
+        address signerOfMessage = ecrecover(messageHash, signature.v, signature.r, signature.s);
+        if (inviter != signerOfMessage) revert UnrecognizedInvitation();
+    }
+
+    function confirmAndAcceptInvitation(address inviter, Signature calldata signature) external {
```
Comment (lucas-janon) [inline packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol:62]: After a user calls `confirmAndAcceptInvitation`, the signature gets broadcasted, becomes public, and anyone could use it to get invited right?

Comment (andresaiello) [reply packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol:62]: yeap

File: packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    function getVerifiedTimestamp(address userAddress) external view returns (uint256) {
+        return userVerificationTimestamps[userAddress];
+    }
+
+    function _verifySignature(address inviter, Signature calldata signature) private pure {
+        bytes32 payloadHash = keccak256(abi.encode(inviter));
+        bytes32 messageHash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", payloadHash));
+
+        address signerOfMessage = ecrecover(messageHash, signature.v, signature.r, signature.s);
```
[the remark this answers is not in the document]
Comment (andresaiello) [reply packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol]: fixed

File: packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    mapping(address => uint256) public userVerificationTimestamps;
+
+    // Records the timestamp when a particular user accepted an invitation from an inviter.
+    mapping(address => mapping(address => uint256)) public acceptedInvitationsTimestamp;
+
+    // Store invitees for each inviter
+    mapping(address => address[]) public inviteeLists;
+
+    // Total invites accepted by day (using the start timestamp of each day as key)
+    mapping(uint256 => uint256) public totalInvitesByDay;
```
Comment (lucas-janon) [inline packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol:21]: Do we really need this? Shouldn't it be on the indexing side?

**Review by lucas-janon — COMMENTED** [review]
I see a potential vulnerability (low) that goes like this:

- I create an invite signature.
- Someone accepts my invite.
- An attacker gets my invite signature since it's public.
- The attacker creates thousands of accounts and gets all of them to be invited by me.
- My `inviteeList` gets extremely long and I get no benefits since they're capped to N invites a week.

I think it's not too bad since we don't iterate over an `inviteeList` on any function, but I think a very long list may cause `view` issues.

On top of that, we have `getInviteeAtIndex`, shouldn't we at least make `index` part of the `InvitationAccepted` event? So there's more information available to query that function.

File: packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol (hunk 3 above)
Comment (andresaiello) [reply packages/zevm-app-contracts/contracts/zeta-points/InvitationManager.sol:21]: we agree with @juansalvatore to keep some data here. the reason is there are two use cases, one is the user page, and the other is the dashboard. for the dashboard we use a graph but for the user maybe we will call directly the contracts

**Comment by andresaiello** [comment]
> I see a potential vulnerability (low) that goes like this:
> 
> * I create an invite signature.
> * Someone accepts my invite.
> * An attacker gets my invite signature since it's public.
> * The attacker creates thousands of accounts and gets all of them to be invited by me.
> * My `inviteeList` gets extremely long and I get no benefits since they're capped to N invites a week.
> 
> I think it's not too bad since we don't iterate over an `inviteeList` on any function, but I think a very long list may cause `view` issues.
> 
> On top of that, we have `getInviteeAtIndex`, shouldn't we at least make `index` part of the `InvitationAccepted` event? So there's more information available to query that function.

that's why we have ` getInviteeAtIndex` and not iterate the list and return the list to avoid any out of gas issue. I agree to add the index, I will do it now

**Review by lucas-janon — APPROVED** [review]
LGTM

**Pull request merged**